Use GitHub Actions Workflow to build on Mac OS X

reznikmm opened the pull request:
See T529-007 for details.

reznikmm commented:
Note: appveyor error is accidental and isn't related to the patch.

raph-amiard commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,66 @@
+          repository: AdaCore/langkit
+          path: langkit
+          fetch-depth: 0
+      - run: pushd langkit; git checkout ${{ github.ref }} || true; popd
```
Could you please:

1. Use `cd` instead of the more obscure `pushd`/`popd` ?
2. Comment this line ? I'm not sure I understand it, even though I guess it's to checkout the corresponding branch ?

TIA

reznikmm replied:
I will. This is a replacement of
```
(
    cd langkit
    python2 ../utils/travis-langkit-branch.py
)
```
It tries to download the same langkit branch as the pushed lal branch (works only for `stable` now).

pmderodat replied:
Hello Maxim,

I’m not sure I understand this part: the `travis-langkit-branch.py` script is designed to select the appropriate langkit branch for all CI situations, the `stable` branch included. So in theory it should be appropriate for GitHub actions. If you have detected a configuration where it does not behave correctly, could you give us the details?

reznikmm replied:
1. I don't have `python2` I guess
2. `travis-langkit-branch.py` uses `TRAVIS_*` variables those not provided by GH

pmderodat replied:
1. The script works fine with Python3: actually the AppVeyor build already runs it with Python3.
2. This should not be a problem: the AppVeyor build just provides the `TRAVIS_*` variables, as renamings of the `APPVEYOR_*` ones (see `utils/appveyor.sh`).

Thanks,

reznikmm replied:
This single line of code meets all my needs for now.

raph-amiard commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,66 @@
+          path: langkit
+          fetch-depth: 0
+      - run: pushd langkit; git checkout ${{ github.ref }} || true; popd
+      - uses: ada-actions/toolchain@b55aaa7
```
Why `@<specific sha1>` instead of `@dev` like written in Fabien's docs ?

reznikmm replied:
`@dev` doesn't include gnat community 2020. There is another branch `ce2020`, but it is unpublished yet on GH marketplace, so SHA1 hash protects us from stealing Bintray API key by Fabien :smiley:

raph-amiard replied:
Ah, ok :) Thanks for the explanation

raph-amiard commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,66 @@
+          export ADALIB_DIR=/tmp/ADALIB_DIR
+          export GPR_PROJECT_PATH=$ADALIB_DIR/share/gpr
+          pip install -r langkit/REQUIREMENTS.dev
+          pip install jsonschema
```
I think if I understand correctly you're using langkit's req file because LAL's is not adapted because it contains a dep on langkit master, right ? I think instead you should:

1. use libadalang's own `REQUIREMENTS.dev`
2. Edit it to remove the dep on langkit, which is annoying 98% of the time
2. Also add the jsonschema dep so you can remove it from here.

Are you OK to do those changes ?

reznikmm replied:
I'm afraid, it's too complicated for me. I just took lines from `utils/travis-install.sh`
```
# Install requirements
pip install -r langkit/REQUIREMENTS.dev
pip install jsonschema
(cd langkit && python setup.py install)

```

raph-amiard replied:
Ok makes sense, I'll take care of the change in that case :)

raph-amiard commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,66 @@
+          pip install jsonschema
+          python langkit/setup.py install
+          make -C gprbuild BUILD=production prefix=$ADALIB_DIR GPRBUILD_OPTIONS="-cargs:C -Dst_mtim=st_mtimespec -gargs" libgpr.build.shared libgpr.install.shared
+          make -C gnatcoll-core PROCESSORS=0 prefix=$ADALIB_DIR LIBRARY_TYPES=relocatable build install
```
Can't you rather `export PROCESSORS=0` at the top ?

raph-amiard commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,66 @@
+          eval `langkit/scripts/manage.sh setenv`
+          ada/manage.py generate -P
+          ada/manage.py build
+          ada/manage.py install $ADALIB_DIR
```
I'm confused, are you building LAL relocatable, and only relocatable ? I thought the ALS needed the static one. Are you sure you're not rebuilding LAL in the ALS's GH actions here ?

raph-amiard replied:
Outdated

reznikmm commented:
Status update since yesterday discussion:
 * I've added steps names as comments
 * replaced `pushd` with `(cd`
 * Added `export PROCESSORS=0`
 * Enabled `static` libraries (both kinds of libraries are build now to help us migrate the ALS to `static` in the feature)
 * DEBUG build doesn't work with `gnatcoll-core` because of error:
```
gnatcoll-vfs.ads:47:01: warning: pragma "ADA_05" ignored
```
 * ` ada/manage.py build --build-mode prod` doesn't work on Mac OS because of error
```
   [link library] libadalang.dylib
Undefined symbols for architecture x86_64:
  "___gmpz_add_ui", referenced from:
      _libadalang__expr_eval__expr_eval_in_env in libadalang-expr_eval.o
```

So I keep just only one working configuration `PROD` gnatcoll-code with `dev` LAL.

raph-amiard commented:
Hey @reznikmm !

Can you try disabling the relocatable build ? It might be what is causing problems in PROD mode.

For the warning, it might mean that gnatcoll's build is failing with GNAT CE 2020, so not a good thing either...

reznikmm commented:
No, I need relocatable libraries to build ALS right now. I would prefer to keep `dev` mode for Libadalang until we change ALS to static build or somebody finds the reason why LAL in `prod` mode doesn't build on MacOs .

reznikmm commented:
I've fixed a few issues in gnatcoll, so now it builds fine in `debug` mode.

reznikmm commented:
Now gh action build 4 archives (prod, dev) x (static x relocatable).
